plugins/lm/lm-nvme.c

```diff
@@ -0,0 +1,894 @@
+		__u8 stype;
+		__u8 seqind;
+		__u8 csuuidi;
+		__u8 csvi;
+		__u8 uidx;
+		__u64 offset;
+		__u32 numd;
+		char  *input;
+		bool  dudmq;
+	};
+
+	struct config cfg = {
+		.sel = -1,
+		.cntlid = 0,
+		.stype = 0,
+		.seqind = 0,
+		.csuuidi = 0,
+		.csvi = 0,
+		.uidx = 0,
+		.offset = 0,
+		.numd = 0,
+		.input = NULL,
+		.dudmq = false
+	};
+
+	OPT_ARGS(opts) = {
+		OPT_BYTE("sel",		's', &cfg.sel, sel),
+		OPT_SHRT("cntlid",	'c', &cfg.cntlid, cntlid),
+		OPT_BYTE("stype",	't', &cfg.stype, stype),
+		OPT_FLAG("dudmq",	'd', &cfg.dudmq, dudmq),
+		OPT_BYTE("seq-ind",	'S', &cfg.seqind, seqind),
+		OPT_BYTE("csuuidi",	'U', &cfg.csuuidi, csuuidi),
+		OPT_BYTE("csvi",	'V', &cfg.csvi, csvi),
+		OPT_BYTE("uidx",	'u', &cfg.uidx, uidx),
+		OPT_LONG("offset",	'o', &cfg.offset, offset),
+		OPT_UINT("numd",	'n', &cfg.numd, numd),
+		OPT_FILE("input-file",	'f', &cfg.input, input),
+		OPT_END()
+	};
+
+	err = parse_and_open(&dev, argc, argv, desc, opts);
+	if (err)
+		return err;
+
+	if (cfg.sel == -1) {
+		nvme_show_error("Select field required");
+		return -EINVAL;
+	}
+
+	// Sanity check input parameters
+	if (cfg.sel == NVME_LM_SEL_SUSPEND || cfg.sel == NVME_LM_SEL_RESUME) {
+		if (cfg.csuuidi != 0 || cfg.csvi != 0) {
+			nvme_show_error("Unexpected fields for %s",
+					ARGSTR(lm_migration_send_select_argstr, cfg.sel));
+			return -EINVAL;
+		}
+	} else if (cfg.sel == NVME_LM_SEL_SET_CONTROLLER_STATE) {
+		if (cfg.dudmq || cfg.stype != 0) {
+			nvme_show_error("Unexpected fields for %s",
+					ARGSTR(lm_migration_send_select_argstr, cfg.sel));
+			return -EINVAL;
+		} else if (!strlen(cfg.input)) {
+			nvme_show_error("Expected file for %s",
+					ARGSTR(lm_migration_send_select_argstr, cfg.sel));
+			return -EINVAL;
+		}
+	}
+
+	if (cfg.input && strlen(cfg.input)) {
+		file = fopen(cfg.input, "r");
+		if (file == NULL) {
+			nvme_show_perror(cfg.input);
+			return -EINVAL;
+		}
+
+		data = nvme_alloc_huge(cfg.numd << 2, &mh);
+		if (!data)
+			return -ENOMEM;
+
+		size_t n_data = fread(data, 1, cfg.numd << 2, file);
+
+		fclose(file);
+
+		if (n_data != (size_t)(cfg.numd << 2)) {
+			nvme_show_error("failed to read controller state data %s", strerror(errno));
+			return -errno;
+		}
+	}
+
+	struct nvme_lm_migration_send_args args = {
+		.args_size = sizeof(args),
+		.fd = dev_fd(dev),
+		.sel = cfg.sel,
+		.mos = NVME_SET(cfg.seqind, LM_MIGRATION_SEND_MOS),
+		.cntlid = cfg.cntlid,
+		.csuuidi = cfg.csuuidi,
+		.uidx = cfg.uidx,
+		.stype = cfg.stype,
+		.offset = cfg.offset,
+		.dudmq = cfg.dudmq,
+		.numd = cfg.numd,
+		.data = data,
+	};
+
+	err = nvme_lm_migration_send(&args);
+	if (err < 0)
+		nvme_show_error("ERROR: nvme_lm_migration_send() failed %s", strerror(errno));
+	else if (err > 0)
+		nvme_show_status(err);
+	else
+		printf("Migration Send (%s) Successful\n",
+		       ARGSTR(lm_migration_send_select_argstr, cfg.sel));
+
+
+	return err;
+}
+
+static void json_controller_state_data(struct nvme_lm_controller_state_data *data, size_t len)
```

Comment:
So it might make sense to support builds without json-c dependencies for this new plugin. Most plugins depend on json-c to be available.  @ikegami-t started to split the plugins so they work also when json-c is missing (see the ocp plugin). I think it would be good to have this from the beginning as it is currently small code base. Though the LM is something in the base spec, so not unlikely to change or get udpated soonish :)

Reply:
Refactored following ocp plugin as an example.